--- a/target/linux/ramips/image/mt7620.mk
+++ b/target/linux/ramips/image/mt7620.mk
@@ -777,7 +777,7 @@ define Device/mercusys_ac12g-v1-8m
   SOC := mt7620a
   IMAGE_SIZE := 7808k
   TPLINK_FLASHLAYOUT := 8Mmtk
-  TPLINK_HWID := 0x04da857c
+  TPLINK_HWID := 0x04f08944
   TPLINK_HWREV := 0x0c000600
   TPLINK_HWREVADD := 0x04000000
   IMAGES += tftp-recovery.bin
@@ -1234,7 +1234,7 @@ define Device/tplink_archer-c5-v4
   SOC := mt7620a
   IMAGE_SIZE := 7808k
   TPLINK_FLASHLAYOUT := 8Mmtk
-  TPLINK_HWID := 0x04da857c
+  TPLINK_HWID := 0x04f08944
   TPLINK_HWREV := 0x0c000600
   TPLINK_HWREVADD := 0x04000000
   IMAGES += tftp-recovery.bin
